extensions/mssql/src/objectExplorerNodeProvider/objectExplorerNodeProvider.ts

```diff
@@ -190,92 +167,105 @@ export class MssqlObjectExplorerNodeProvider extends ProviderBase implements sql
+	async findSqlClusterNodeBySqlContext<T extends TreeNode>(sqlContext: ICommandObjectExplorerContext | sqlops.ObjectExplorerContext): Promise<T> {
 		let node: T = undefined;
-		let session = await this.findSessionForConnection(explorerContext.connectionProfile);
+		let sqlOeContext = 'explorerContext' in sqlContext ? sqlContext.explorerContext : sqlContext;
+		let sqlConnProfile = sqlOeContext.connectionProfile;
+		let session = this.findSqlClusterSessionBySqlConnProfile(sqlConnProfile);
 		if (session) {
-			if (explorerContext.isConnectionNode) {
+			if (sqlOeContext.isConnectionNode) {
 				// Note: ideally fix so we verify T matches RootNode and go from there
-				node = <T><any>session.root;
+				node = <T><any>session.rootNode;
 			} else {
 				// Find the node under the session
-				node = <T><any>await session.root.findNodeByPath(explorerContext.nodeInfo.nodePath, true);
+				node = <T><any>await session.rootNode.findNodeByPath(sqlOeContext.nodeInfo.nodePath, true);
 			}
 		}
 		return node;
 	}
 
-	private async findSessionForConnection(connectionProfile: sqlops.IConnectionProfile): Promise<Session> {
-		for (let session of this.sessionMap.values()) {
-			if (session.connection && await session.connection.isMatch(connectionProfile)) {
-				return session;
-			}
-		}
-		return undefined;
+	public findSqlClusterSessionBySqlConnProfile(sqlConnProfile: sqlops.IConnectionProfile): SqlClusterSession {
+		return Array.from(this.sessionMap.values())
+			.find(s => s.isMatchedSqlConnProfile(sqlConnProfile));
 	}
 }
 
-export class Session {
-	private _root: RootNode;
-	constructor(private _connection: Connection, private sessionId?: string) {
-	}
+export class SqlClusterSession {
+	private _sqlClusterConnection: SqlClusterConnection;
+	private _sqlSession: sqlops.ObjectExplorerSession;
+	private _sqlConnProfile: sqlops.IConnectionProfile;
+	private _rootNode: SqlClusterRootNode;
 
-	public get uri(): string {
-		return this.sessionId || this._connection.uri;
+	constructor(arg: SqlClusterSessionArg) {
+		this._sqlClusterConnection = arg.sqlClusterConnection;
+		this._sqlSession = arg.sqlSession;
+		this._sqlConnProfile = arg.sqlConnProfile;
+		this._rootNode = new SqlClusterRootNode(this,
+			new TreeDataContext(arg.appContext.extensionContext, arg.changeHandler),
+			arg.sqlSession.rootNode.nodePath);
 	}
 
-	public get connection(): Connection {
-		return this._connection;
-	}
+	public get sqlClusterConnection(): SqlClusterConnection { return this._sqlClusterConnection; }
+	public get sqlSession(): sqlops.ObjectExplorerSession { return this._sqlSession; }
+	public get sqlConnProfile(): sqlops.IConnectionProfile { return this._sqlConnProfile; }
+	public get sessionId(): string { return this._sqlSession.sessionId; }
+	public get rootNode(): SqlClusterRootNode { return this._rootNode; }
 
-	public set root(node: RootNode) {
-		this._root = node;
+	public isMatchedSqlConnProfile(sqlConnProfile: sqlops.IConnectionProfile): boolean {
```

Comment:
`isMatchedSqlConnection` perhaps? Or maybe use TSql to differentiate between SQL and the Knox / SQLCluster connectin?

Reply:
fixed.